cdc/log.cc

```diff
@@ -880,12 +881,24 @@ class log_mutation_builder {
         return _base_schema;
     }
 
+    clustering_key create_ck(int batch) const {
+        auto log_ck = clustering_key::from_exploded(_log_schema, { _tuuid, int32_type->decompose(batch) });
+        set_key_columns(log_ck, _base_schema.partition_key_columns(), _base_pk);
+        return log_ck;
+    }
+
     // Creates a new clustering row in the mutation, assigning it the next `cdc$batch_seq_no`.
     // The numbering of batch sequence numbers starts from 0.
     clustering_key allocate_new_log_row() {
-        auto log_ck = clustering_key::from_exploded(_log_schema, { _tuuid, int32_type->decompose(_batch_no++) });
-        set_key_columns(log_ck, _base_schema.partition_key_columns(), _base_pk);
-        return log_ck;
+        return create_ck(_batch_no++);
+    }
+
+    bool has_rows() const {
+        return _batch_no != 0;
+    }
+
+    clustering_key last_row_key() const {
```

Comment:
One would presume that this is a pure calculation, but it actually modifies the state now.

Reply:
Fixed.